### ISSUE TITLE
Cannot receive SMS messages

### ISSUE THREAD

**Issue by GoogleCodeExporter** [msg4real/pygooglevoice#6]
```
pygooglevoice is currently not providing an API for receiving SMS messages. I 
looked to see why, 
and quickly figured it out: Google is not providing any easy ways to access 
them.

However, they can be accessed, and I have been for the last few months 
successfully.

What I needed was something that would actually pop a notification and sound an 
alarm when 
someone texted me, like in an instant messenger. I can't check google.com/voice 
every 30 
seconds. It disrupts whatever I'm doing.

This is my first Python project, so bear with me if the code is a little odd or 
misshapen. :)

First, I've extended the original Voice object from pygooglevoice to create two 
new functions, 
called "get_sms_data (self, smsID)" and "set_read_status (self, smsID, 
readStatus)". smsID refers 
to a particular conversation "bubble" in google.com/voice.

get_sms_data calls Google's "inbox/recent/sms/" Javascript API, and passes the 
resulting raw 
HTML data to a new class I wrote titled GVoiceSMSProcessor. This class is 
called up as follows:

smsProcessor = GVoiceSMSProcessor(smsData, smsID)

Internally, GVoiceSMSProcessor organizes all the available data into a 
dictionary called 
conversations. Here is a sample dump of smsProcessor.conversations, with 
personal data 
removed, so you get an idea of how awesome this is:

[
{'contact': 'PersonA',
'messages':
[
{'text': 'Hey', 'sender': 'PersonA', 'time': '5:09 PM'},
{'text': "Yea, what's up?", 'sender': 'Me', 'time': '5:09 PM'},
{'text': "I'm at <place> if u want to come over here and hang out.", 'sender': 
'PersonA', 'time': 
'5:15 PM'},
]
}]

I could've easily had it process more data, but I included here only the data 
that was not 
retrieved in pygooglevoice's API.

The GVoiceSMSProcessor object is somewhat confusing to read through, but the 
idea is quite 
simple: process the raw data with BeautifulSoup, and then loop through the 
entire document, 
element by element. The code in "__searchHTML" looks for specific class 
patterns in the HTML. 
When it finds a tag that contains a specific piece of information, it logs what 
that data is we are 
so that farther down it can be processed into the dictionary. The 
'contact-next' state occurs 
when the contact name is "Me", because the tag that marks when the contact name 
is coming up 
is empty, and it's the next (classless) tag that has the good stuff. Thus, this 
search method also 
tracks a little on state.

If you look at the contents of https://www.google.com/voice/inbox/recent/sms/ 
and find each 
class, you'll see why the searching does what it does (hopefully).

I'm attaching the entire client I wrote for myself, so other users can try it. 
Linux users, "growl" 
ought to be changed to some other notification daemon. Also, mplayer is used to 
play a 
notification sound, which is specific to OSX.

In summary, the pygooglevoice developers should look to the gvoice.py to see 
what additions 
can be made to their code code.
```

Original issue reported on code.google.com by `jacobgod...@gmail.com` on 14 Sep 2009 at 10:46

Attachments:
- [pygooglevoice-sms.tar.bz2](https://storage.googleapis.com/google-code-attachments/pygooglevoice/issue-6/comment-0/pygooglevoice-sms.tar.bz2)
